_posts/2019-01-26-generic-way-of-handling-loading-saving-and-validationerrors-in-angular.md

```diff
@@ -0,0 +1,410 @@
+
+We can now use the `HttpStatusService` in our code as easy as this:
+
+```typescript
+export class UserComponent {
+  loading$ = this.httpStatus.loading;
+  validationErrors$ = this.httpStatus.validationErrors;
+  acting$ = this.httpStatus.loading;
+
+  constructor(
+    private httpStatusService: HttpStatusService) {
+  }
+}
+```
+
+We now have 3 observables that can easily be consumed in the template of the component with the use of the [async](https://angular.io/api/common/AsyncPipe) pipe. Here is an example:
+
+```html
+<my-spinner *ngIf="loading$|async"></my-spinner>
+<my-user-form 
+    [validationErrors]="validationErrors$|async"
+    [disabled]="acting$|async"></my-user-form>
+
+```
+
+## Optimizing with decorators
+
+We have cleaned up a lot, we found an easy way to get the httpstatusses to the component, but we can make it even simpler with the use of decorators. let's refactor the `UserComponent` class accordingly:
+
+```typescript
+export class UserComponent {
+  @Loading()loading$;
+  @ValidationErrors() validationErrors$;
+  @Acting() acting$;
+}
+```
+
+This is very declarative way of working. The component is way cleaner and we don't have to inject the `HttpStatusService` anymore.
+
+But how do we create these decorators? I'm glad you asked, it's pretty easy. A property decorator is simply a function that returns a function that gets the target and key as arguments.
+
+```typescript
+export function Loading() {
+  return function (target: any, key: string): void {
+    // in this case the target is the component
+    // instance and key the property name
+    // target: userComponent, key: loading$
+    // now we have to set the property to the actual 
+    // loading$ observable that lives in the HttpStatusService
+    target[key] = // todo
+  }
+};
+```
+
+To set the property of the target we need to inject the `HttpStatusService` instance that is registered on the root injector of our application. After all that's the instance that contains the actual state. Currently there is no easy way to do that.
+Until Angular provides us with functionality like [that](https://github.com/angular/angular/issues/23301) we can use the following solution:
+
+Next to the `http-status.service` file, create a file called `root-injector.ts` and add the following code:
+
+```typescript
+import { Injector } from '@angular/core';
+
+export let rootInjector: Injector;
+export function setRootInjector(injector: Injector): void {
+  rootInjector = injector;
+}
+```
+
+The `setRootInjector()` function will be used by the rootModule to set the `rootInjector` variable that we expose here.
+To make it work the rootModule will have to call the `setRootInjector()` function like this:
+
+```typescript
+@NgModule({
+    ...
+})
+export class AppModule {
+  constructor(private injector: Injector) {
+    setRootInjector(injector);
+  }
+}
+
+```
+
+The last step is to actually use the `rootInjector` variable inside the decorators. The result looks like this:
```

Comment:
Why not decorate the class with a custom class decorator to inject the service instead?

Reply:
You still cant access the injector there, You cant fetch a ComponentRef from a component without injecting stuff there, do you have an example of a working solution?